src/stm32_generic/i2c.cpp

```diff
@@ -0,0 +1,548 @@
+  /// High level (Write 1 to GPIOx_ODR).
+  sda->configure({ .resistor = pin_resistor::none, .open_drain = true });
+  scl->configure({ .resistor = pin_resistor::none, .open_drain = true });
+  sda->level(true);
+  scl->level(true);
+  // /// 3. Check SCL and SDA High level in GPIOx_IDR.
+  // while (!sda->level() || !scl->level()) {
+  //   continue;
+  // }
+  // /// 4. Configure the SDA I/O as General Purpose Output Open-Drain, Low
+  // level
+  // /// (Write 0 to GPIOx_ODR).
+  // sda->level(false);
+  // scl->level(false);
+  // /// 5. Check SDA Low level in GPIOx_IDR.
+  // while (sda->level() || scl->level()) {
+  //   continue;
+  // }
+  // /// 8. Configure the SCL I/O as General Purpose Output Open-Drain, High
+  // level
+  // /// (Write 1 to GPIOx_ODR).
+  // sda->level(true);
+  // scl->level(true);
+
+  // /// 9. Check SCL High level in GPIOx_IDR.
+  // while (!sda->level() || !scl->level()) {
+  //   continue;
+  // }
+
+  /// 13. Set SWRST bit in I2Cx_CR1 register.
+  // bit_modify(i2c_reg->cr1).set(i2c_cr1::software_reset);
+
+  // /// 14. Clear SWRST bit in I2Cx_CR1 register.
+  // bit_modify(i2c_reg->cr1).clear(i2c_cr1::software_reset);
+
+  // /// 15. Enable the I2C peripheral by setting the PE bit in I2Cx_CR1
+  // register. bit_modify(i2c_reg->cr1).set(i2c_cr1::peripheral_enable);
+
+  // Set the SDA line to high, so that after the slave releases
+  // the SDA and listen for ACK/NACK, it will get a NACK.
+  sda->level(true);
+
+  for (int i = 0; i < 10; ++i) {
+    // Set SCL high.
+    scl->level(true);
+    continue;
+    // Set SCL low.
+    scl->level(false);
+    continue;
+  }
+  // Set SCL high. When I2C is idle the line should be pulled high.
+  scl->level(true);
+}
+
+void i2c::configure(hal::i2c::settings const& p_settings, hertz p_frequency)
+{
+  constexpr auto slow_mode_max_speed = 100_Hz;
+  if (static_cast<u32>(p_frequency) < 2_MHz ||
+      static_cast<u32>(p_frequency) > 50_MHz) {
+    safe_throw(hal::argument_out_of_domain(this));
+  }
+  auto i2c_reg = get_i2c_reg(m_i2c);
+  bit_modify(i2c_reg->cr1).set(i2c_cr1::software_reset);
+  bit_modify(i2c_reg->cr1).clear(i2c_cr1::software_reset);
+
+  bit_modify(i2c_reg->filter)
+    .set(i2c_filter::digital_filter)
+    .clear(i2c_filter::analog);
+
+  /// I2C communication speed, fSCL ~ 1/(thigh + tlow). The real frequency may
+  /// differ due to the analog noise filter input delay.
+  /// CH 18.6.8 in RM0383 (stm32f411 user manual)
+
+  u16 const ccr_value = static_cast<hal::u32>(p_frequency) /
+                        (2 * static_cast<hal::u32>(p_settings.clock_rate));
+  if (p_settings.clock_rate > slow_mode_max_speed) {
+    if (ccr_value > 4) {
+      bit_modify(i2c_reg->ccr)
+        .set(i2c_ccr::i2c_mode_sel)
+        .insert(i2c_ccr::ccr, ccr_value);
+    } else {
+      safe_throw(hal::argument_out_of_domain(this));
+    }
+  } else {
+    bit_modify(i2c_reg->ccr)
+      .clear(i2c_ccr::i2c_mode_sel)
+      .insert(i2c_ccr::ccr, ccr_value);
```

Comment:
Same here, with using the template argument. Prefer the template arugment unless the template version cannot work because the mask is only known at runtime.